srt-protocol/src/protocol/sender/mod.rs

```diff
@@ -163,66 +146,74 @@ impl Sender {
-        if from != self.settings.remote {
-            return;
-        }
-
-        debug!("Received packet {:?}", packet);
-
-        match packet {
-            Packet::Control(control) => self.handle_control_packet(control, now),
-            Packet::Data(_) => {}
-        }
-    }
-
     pub fn is_flushed(&self) -> bool {
-        trace!("{:?} Checking is flushed: ll.len()={}, tb.len()={}, lrap={}, nsn={}, sb.len()={}, ob.len()={}", self.settings.local_sockid, self.loss_list.len(), 
-            self.transmit_buffer.len(), self.lr_acked_packet, self.transmit_buffer.next_sequence_number, self.send_buffer.len(), self.output_buffer.len());
+        debug!(
+            "{:?}|{:?}|recv - ll.len()={}, tb.len()={}, lrap={}, nsn={}, sb.len()={}, ob.len()={}",
+            TimeSpan::from_interval(self.settings.socket_start_time, Instant::now()),
+            self.settings.local_sockid,
+            self.loss_list.len(),
+            self.transmit_buffer.len(),
+            self.lr_acked_packet,
+            self.transmit_buffer.next_sequence_number,
+            self.send_buffer.len(),
+            self.output_buffer.len()
+        );
         self.loss_list.is_empty()
             && self.transmit_buffer.is_empty()
             && self.lr_acked_packet == self.transmit_buffer.next_sequence_number
             && self.output_buffer.is_empty()
     }
 
-    pub fn pop_output(&mut self) -> Option<(Packet, SocketAddr)> {
+    pub fn next_packet(&mut self) -> Option<(Packet, SocketAddr)> {
         let to = self.settings.remote;
         self.output_buffer
             .pop_front()
             .map(move |packet| (packet, to))
     }
 
-    pub fn next_action(&mut self, now: Instant) -> SenderAlgorithmAction {
-        use SenderAlgorithmAction::*;
-        use SenderAlgorithmStep::*;
-
-        // don't return close until fully flushed
-        if self.close_requested && self.is_flushed() {
-            if !self.shutdown_sent {
-                debug!("{:?} sending shutdown", self.settings.local_sockid);
-                self.send_control(ControlTypes::Shutdown, now);
-                self.shutdown_sent = true;
-            }
-            return Close;
-        }
-
+    pub fn check_timers(&mut self, now: Instant) {
         if let Some(exp_time) = self.snd_timer.check_expired(now) {
             self.on_snd_event(exp_time);
         }
         if let Some(exp_time) = self.keepalive_timer.check_expired(now) {
             self.on_keepalive_event(exp_time);
         }
 
-        if self.step == Step6 {
-            return WaitUntil(self.snd_timer.next_instant());
+        // don't return close until fully flushed
+        let flushed = self.is_flushed();
+        if self.status.check_shutdown(now, || flushed) {
+            debug!("{:?} sending shutdown", self.settings.local_sockid);
+            self.send_control(ControlTypes::Shutdown, now);
         }
+    }
+
+    pub fn next_timer(&self, now: Instant) -> Instant {
```

Comment:
I think this here is why the sender algorithm now wakes up so much more than it used to: even when there is no data to be sent it requests wakeup to send a packet!

Reply:
Good point! :)

Reply:
My first pass at fixing this just now reduces the execution time for timestamp_rollover to roughly 7min on my MacBook. I'm still not pleased with this. But it that is promising, so I'll run flamegaph on it and see if there are any low hanging fruit for further optimization.

Reply:
Sounds good! We can probably just run 3 hours of simulation instead of a full 24, as rollover occurs every ~1.2 hours, which should help by almost an order of magnitude.